File: lib/ruby/intcode/op_code.rb
# frozen_string_literal: true
require 'scanf'

class OpCode
  ADD  = 1
  MULT = 2
  IN   = 3
  OUT  = 4
  JNZ  = 5
  JZ   = 6
  STOL = 7
  STOE = 8
  HALT = 99

  attr_accessor :opcode, :modes, :size

  def initialize(opcode)
    @opcode, @modes = deconstruct_opcode(opcode)
    @size = op_size(@opcode)
  end

  def exec(ip, mem)
    case self
    when ADD  then add_op(ip, mem)
    when MULT then mult_op(ip, mem)
    when IN   then input(ip, mem)
    when OUT  then output(ip, mem)
    when JNZ  then jump_if_true(ip, mem)
    when JZ   then jump_if_false(ip, mem)
    when STOL then store_if_less(ip, mem)
    when STOE then store_if_equal(ip, mem)
    end
  end

  def ==(other)
    case other
    when OpCode
      opcode == other.opcode
    when Integer
      opcode == other
    end
  end

  def ===(other)
    case other
    when OpCode
      opcode === other.opcode
    when Integer
      opcode === other
    end
  end

  private

  def deconstruct_opcode(code)
    opcode = code % 100
    modes = enum_modes(code)
    [opcode, modes]
  end

  def enum_modes(code)
    digits = (code / 100).digits
    size = digits.size
    digits.fill(0, size, 3 - size) if size < 3
    digits.collect do |i|
      case i
      when 0 then :mem
      when 1 then :imm
      end
    end
  end

  def op_size(code)
    case code
    when ADD, MULT,
         STOL, STOE then 4
    when JZ, JNZ    then 3
    when IN, OUT    then 2
    when HALT       then 1
    end
  end

  def read_operands(ip, mem)
    mem[ip + 1, size - 1]
  end

  def add_op(ip, mem)
    modes[2] = :imm
    val1, val2, addr = load_operands(ip, mem)
    mem[addr] = val1 + val2
    ip + size
  end

  def mult_op(ip, mem)
    modes[2] = :imm
    val1, val2, addr = load_operands(ip, mem)
    mem[addr] = val1 * val2
    ip + size
  end

  def input(ip, mem)
    print '? '
    val = scanf('%d').first
    operand = read_operands(ip, mem).first
    mem[operand] = val
    ip + size
  end

  def output(ip, mem)
    puts load_operands(ip, mem).first
    ip + size
  end

  def jump_if_true(ip, mem)
    val, addr = load_operands(ip, mem)
    val.zero? ? ip + size : addr
  end

  def jump_if_false(ip, mem)
    val, addr = load_operands(ip, mem)
    val.zero? ? addr : ip + size
  end

  def store_if_less(ip, mem)
    modes[2] = :imm
    val1, val2, addr = load_operands(ip, mem)
    mem[addr] = val1 < val2 ? 1 : 0
    ip + size
  end

  def store_if_equal(ip, mem)
    modes[2] = :imm
    val1, val2, addr = load_operands(ip, mem)
    mem[addr] = val1 == val2 ? 1 : 0
    ip + size
  end

  def load_operands(ip, mem)
    read_operands(ip, mem).zip(modes).collect do |val, mode|
      mode == :mem ? mem[val] : val
    end
  end
end

File: lib/rust/intcode/src/op_code.rs
use std::io::{stdin,stdout};
use std::io::Write;

#[derive(Clone,Copy,Debug,Eq,PartialEq)]
pub enum Mneumonic {
    Add,
    Mult,
    In,
    Out,
    Jnz,
    Jz,
    Stol,
    Stoe,
    Halt
}

#[derive(Clone,Copy,Debug,Eq,PartialEq)]
pub enum Mode { Mem, Imm }

pub struct OpCode {
    mne: Mneumonic,
    modes: [Mode; 3],
    size: u8
}

impl OpCode {
    pub fn exec(&self, ip: usize, mem: &mut [isize]) -> usize {
        use Mneumonic::*;

        match self.mne {
            Add  => Self::add_op(self, ip, mem),
            Mult => Self::mult_op(self, ip, mem),
            In   => Self::input(self, ip, mem),
            Out  => Self::output(self, ip, mem),
            Jnz  => Self::jump_if_true(self, ip, mem),
            Jz   => Self::jump_if_false(self, ip, mem),
            Stol => Self::store_if_less(self, ip, mem),
            Stoe => Self::store_if_equal(self, ip, mem),
            _    => 0
        }
    }

    #[inline]
    fn add_op(&self, ip: usize, mem: &mut [isize]) -> usize {
        let o = Self::load_operands(self, ip, mem);
        let addr = o[2] as usize;
        mem[addr] = o[0] + o[1];
        ip + self.size as usize
    }

    #[inline]
    fn mult_op(&self, ip: usize, mem: &mut [isize]) -> usize {
        let o = Self::load_operands(self, ip, mem);
        let addr = o[2] as usize;
        mem[addr] = o[0] * o[1];
        ip + self.size as usize
    }

    #[inline]
    fn input(&self, ip: usize, mem: &mut [isize]) -> usize {
        print!("? ");
        stdout().flush().unwrap();

        let o = Self::read_operands(self, ip, mem);
        let addr = o[0] as usize;
        let mut buf = String::new();
        stdin().read_line(&mut buf).unwrap();
        mem[addr] = buf.trim().parse().unwrap();
        ip + self.size as usize
    }

    #[inline]
    fn output(&self, ip: usize, mem: &[isize]) -> usize {
        let o = Self::load_operands(self, ip, mem);
        println!("{}", o[0]);
        ip + self.size as usize
    }

    #[inline]
    fn jump_if_true(&self, ip: usize, mem: &[isize]) -> usize {
        let o = Self::load_operands(self, ip, mem);
        if o[0] == 0 { ip + self.size as usize } else { o[1] as usize }
    }

    #[inline]
    fn jump_if_false(&self, ip: usize, mem: &[isize]) -> usize {
        let o = Self::load_operands(self, ip, mem);
        if o[0] == 0 { o[1] as usize } else { ip + self.size as usize }
    }

    #[inline]
    fn store_if_less(&self, ip: usize, mem: &mut [isize]) -> usize {
        let o = Self::load_operands(self, ip, mem);
        let addr = o[2] as usize;
        mem[addr] = if o[0] < o[1] { 1 } else { 0 };
        ip + self.size as usize
    }

    #[inline]
    fn store_if_equal(&self, ip: usize, mem: &mut [isize]) -> usize {
        let o = Self::load_operands(self, ip, mem);
        let addr = o[2] as usize;
        mem[addr] = if o[0] == o[1] { 1 } else { 0 };
        ip + self.size as usize
    }

    #[inline]
    fn read_operands<'t>(&self, ip: usize, mem: &'t [isize]) -> &'t [isize] {
        &mem[ip + 1 .. (ip + self.size as usize)]
    }

    #[inline]
    fn load_operands<'t>(&self, ip: usize, mem: &'t [isize]) -> Vec<isize> {
        use Mode::Mem;
        Self::read_operands(self, ip, mem).iter()
            .zip(&self.modes)
            .map(|(val, mode)| if *mode == Mem { mem[*val as usize] } else { *val })
            .collect()
    }
}

impl From<isize> for OpCode {
    fn from(item: isize) -> Self {
        use Mode::Imm;
        use Mneumonic::*;

        let code = item % 100;
        let mut modes = read_modes(item / 100);
        match code {
            1  => {
                modes[2] = Imm;
                OpCode { mne: Add,  modes, size: 4 }
            },

            2  => {
                modes[2] = Imm;
                OpCode { mne: Mult, modes, size: 4 }
            },

            3  => OpCode { mne: In,   modes, size: 2 },
            4  => OpCode { mne: Out,  modes, size: 2 },
            5  => OpCode { mne: Jnz,  modes, size: 3 },
            6  => OpCode { mne: Jz,   modes, size: 3 },

            7  => {
                modes[2] = Imm;
                OpCode { mne: Stol, modes, size: 4 }
            }

            8  => {
                modes[2] = Imm;
                OpCode { mne: Stoe, modes, size: 4 }
            },

            99 => OpCode { mne: Halt, modes, size: 1 },
            _ => panic!("Invalid OpCode")
        }
    }
}

fn read_modes(code: isize) -> [Mode; 3] {
    use Mode::*;

    let mut i = 0;
    let mut code = code;
    let mut modes = [Mem; 3];
    while code > 0 {
        let mode = if code % 10 == 1 { Imm } else { Mem };
        modes[i] = mode;
        i += 1;
        code /= 10;
    }
    modes
}

impl Eq for OpCode { }

impl PartialEq for OpCode {
    fn eq(&self, other: &Self) -> bool {
        self.mne == other.mne
    }
}

impl PartialEq<Mneumonic> for OpCode {
    fn eq(&self, other: &Mneumonic) -> bool {
        self.mne == *other
    }
}
